core/dbt/contracts/graph/nodes.py

```diff
@@ -168,6 +168,12 @@ class ColumnLevelConstraint(dbtClassMixin):
     warn_unsupported: bool = (
         True  # Warn if constraint is not supported by the platform and won't be in DDL
     )
+    columns: List[str] = field(default_factory=list)
+
+
+@dataclass
+class ModelLevelConstraint(ColumnLevelConstraint):
+    pass  # REVIEW: The columns property above should be on this class instead
```

Comment:
For reasons I am still working to understand, moving the columns property above to this class, where it would actually makes sense, results in the columns property being missing when the raw dictionary is passed to the adapter implementation of render_raw_model_constraint().

@gerda It seems like there is a mechanism at work which adds defaults values to and removes unexpected values from the raw dictionaries, but is confused about which type this one has. Can you help me understand what's going on here?

Reply:
"constraints" is not processed as a config, it's a node-level field. So you need to actually copy it at the top of the "patch" method on parsed node, where "description" and "columns" and "access" are copied. It's only on model nodes so you'd also want to check the node.resource_type. (At some point we may want to split out more finely grained "patch" methods...)

Reply:
I couldn't find the answer, but I want to make sure I actually understand the question. It sounds like if `columns` is a key in `raw_constraint`, then `BaseAdapter._parse_model_constraint(raw_constraint)` does not have a `columns` attribute, but `BaseAdapter._parse_column_constraint(raw_constraint)` does have a `columns` attribute. Is that right?

Is `raw_constraint["columns"]` just a `List[str]`? Or is it more complicated?

Reply:
>  (At some point we may want to split out more finely grained "patch" methods...)

Thinking along the same lines for patching other new node-level fields that are only on model nodes [in the model version parsing spike (version, is_latest_version)](https://github.com/dbt-labs/dbt-core/pull/7204/files#diff-42cfc99b2fcd89ab75957b8999bee2b1af822f7d39953ca99f1f285c170557c1R396-R399).